src/cuda/api/array.hpp

```diff
@@ -0,0 +1,153 @@
+/**
+ * @file array.hpp
+ * 
+ * @brief A CUDA array wrapper class. Especially useful
+ * if used in combination with CUDA textures.
+ *
+ */
+
+#pragma once
+#ifndef CUDA_API_WRAPPERS_ARRAY_HPP
+#define CUDA_API_WRAPPERS_ARRAY_HPP
+
+#include <cuda/api/current_device.hpp>
+#include <cuda/api/error.hpp>
+
+#include <cuda_runtime.h>
+#include <cuda_runtime_api.h>
+
+#include <algorithm>
+#include <array>
+#include <numeric>
+
+namespace cuda {
+
+template<bool AssumedCurrent> class device_t;
+
+namespace array {
+
+namespace detail {
+
+/**
+ * @brief Do not use directly
+ *
+ */
+template <typename T, size_t NDIMS>
+class array_base {
```

Comment:
Have you considered the possibility of having just a single templated `array_t` class, and then having some of the methods specialized by dimension, and other methods being common?

Reply:
Just implemented this; see Commit a5779c3.